Adding support for telescope containers and ties in astrology.lic

jayRyan24 opened the pull request:
I will add support for walkingastro as well. 

Settings:

```
telescope_storage:
  tied: toolstrap
```
or 
```
telescope_storage:
  container: backpack
```

If both are listed, it will tie before using the container. 

Changes for #3763.

jayRyan24 commented:
https://pastebin.com/CXRJPVsN

jayRyan24 commented on profiles/base.yaml:
```diff
@@ -1120,6 +1120,7 @@ held_athletics_items:
 # https://elanthipedia.play.net/Lich_script_repository#astrology
 astrology_buffs:
 have_telescope: false
+telescope_storage:
```
Do I need this if I also have it in base-empty, I can't remember.

rpherbig replied:
Yes

Kaesken commented:
I updated everything and added the tied as stated above but it keeps wanting to use "get my telescope" instead of untying it from my toolbelt.

rpherbig commented on common-moonmage.lic:
```diff
@@ -12,8 +12,24 @@ module DRCMM
 
-  def get_telescope
-    DRC.bput('get my telescope', 'you get', 'You are already holding that.', "That can't be picked up", 'You need a free hand to pick that up.', 'You should stop practicing your Athletics skill before you do that')
+  def get_telescope(storage)
```
❌  Could you pull the `common-moonmage` changes into a separate PR? We want to merge it before the other PR to give people time to get the update.

jayRyan24 replied:
Yeah, will do that.

rpherbig commented on profiles/base-empty.yaml:
```diff
@@ -67,3 +67,4 @@ empty_values:
   held_athletics_items: []
   training_list: []
   stabbity: {}
+  telescope_storage: []
```
❓ Shouldn't this be a hash? `telescope_storage: {}`

jayRyan24 replied:
Yes

jayRyan24 commented:
> I updated everything and added the tied as stated above but it keeps wanting to use "get my telescope" instead of untying it from my toolbelt.

This is probably in reference to what @rpherbig is talking about. You would need to pull in the common-moonmage changes, and then run ;common-moonmage before any of the other stuff will work.

rpherbig commented:
✔️ Looks good. If I haven't merged this in a few days, remind me.

jayRyan24 commented:
> ✔️ Looks good. If I haven't merged this in a few days, remind me.

Will do.

jayRyan24 commented:
@rpherbig I did add changes to the file just now. Basically, Astrology will constantly get & stow the telescope each iteration. I moved that logic out (now that it's methodized), before and after the observe logic.

rpherbig commented on astrology.lic:
```diff
@@ -168,7 +169,7 @@ class Astrology
     perc_time = Time.now - 61
     Flags.add('rtr-expire', get_data('spells').spell_data['Read the Ripples']['expire']) if rtr_active?
-    DRCMM.get_telescope(@tele_storage) if @have_telescope
+    get_telescope(@tele_storage) if @have_telescope
```
💡 I like not having the prefix so the code is concise, but I've recently decided that, overall, I prefer to have it. It makes it easier for new contributors to figure out where helpers come from.

Not required for you to change this, just getting my thoughts out there.

jayRyan24 replied:
My plan is to make DRCMM's methods more robust and add them to astrology.lic. 

```
if @have_telescope
      case center_telescope_on(target)
      when /open it/
```

jayRyan24 replied:
I will do whatever you want me to do in this situation, just let me know. :)

jayRyan24 replied:
```
  def roll_bones(is_tied, container)
    if get_bones
      DRC.bput('roll my bones', 'roundtime')
      store_bones
    end
  end
```

rpherbig replied:
I love the plan.

As for the prefix, if you come back to this file for some reason please add it, no need to bother with another commit just for that change.

jayRyan24 replied:
I'm just waiting for all of these things to get merged, then I will update the whole.

jayRyan24 replied:
I reverted the changes as I would just be doing that in the next PR. No worries!

Kaesken commented:
Maybe I'm not understanding?  I replaced each of the files with what's in the file change tab.

Log:  https://pastebin.com/zAqG5P8V
YAML:  https://pastebin.com/P9KtnWZw

jayRyan24 commented:
> Maybe I'm not understanding? I replaced each of the files with what's in the file change tab.
> 
> Log: https://pastebin.com/zAqG5P8V
> YAML: https://pastebin.com/P9KtnWZw
> 
> Even tried killing common-moonmage and restarting it but it won't tell me whether or not it's been started and wont' do anything for it so I run training-manager. How can I get common-moonmage to work by itself? I apologize in advance for any difficulty on my part, lol. This section of github is really new to me.

Hey, sorry you're having so much trouble. So basically, you would need to start up the system and let everything load

`,kill walkingastro`

Due to how the file system works, whenever you run dependency, it will update all the files and bring in new changes. This can be stopped with developer mode, but it's not something you should do unless you know how to handle it. What you can do is run the system like normal, and then make all the changes to the files with copy and paste. 

Then:
```
,common-moonmage
,walkingastro
```

This should then start up and let you test. But once you restart dependency, i.e. log out and back in, or even `,k dependency` and `,dependency`, it will overwrite whatever changes you had made.

Kaesken commented:
Oh, no need to be sorry, hun!  I am in developer mode because I do edit a few of the files but they're just a few tweaks.  I can get off dev mode and have dependency update everything then go back and remake my changed and see if it'll work.  I didn't even think of that.  I didn't restart dependency, tho.  Going to try all that now.  :)

jayRyan24 commented:
You'll lose your changes. If in dev mode, you should already be fine. You would just need to copy and paste all those files. But it did go through, so you'll get the changes there I guess, lol.

jayRyan24 commented:
@rpherbig Here's the log with the get/stow changes. A lot more clean as we aren't getting and stowing each observe. 

https://pastebin.com/qi6r0V6k

Kaesken commented:
walkingastro wasn't running, but then after I killed dependency and restarted it, this is what I got when starting walkingastro:  https://pastebin.com/6AzCe4Gf

This is what I get when trying to stop/start common-moonmage:  https://pastebin.com/qSzVVfi9
I can't even tell if it does anything or what it does.

I also forgot to mention that I was trying to run "training-manager" with the telescope.  I have it in settings for her to do astro in her hunting which isn't untying it which it still tries to "get my telescope" in the combat-trainer.

[combat-trainer]>get my telescope
You should untie the e'erdream telescope from the teller's toolbelt first.

I completely forgot to check if that was possible or if you guys were just working on the other files first.  My apologies on that.

But!  The walking astro one works!  Just the match doesn't match for me, hehe

Kaesken commented:
Ok, so I got my walkingastro to finally work with untying the telescope!!

I had to add 'You untie' to this part under walkingastro.lic
  def get_telescope
    case DRCMM.get_telescope(@tele_storage)
    when 'You get', 'You remove', 'You untie'

And change 'you untie your' to 'You untie' in common-moonmage and it's working perfectly for walkingastro.

jayRyan24 commented:
OK, thanks a lot. I didn't have a tie to test with!

jayRyan24 commented:
@rpherbig Can we get this merged? I can handle the questions (and enough people are using the change) that we can just instruct people to restart dep to pick up the changes. 

I am kind of in the mood to get the changes for the common file and astrology moving further.

Tarhish commented:
The only reason I'm aware of is that some people run multiple characters out of the same lich setup.  Certain common files don't get updated if they're being used at the time you login, so there have been issues in the past with everything crashing and burning because someone got one script updated but not the common file it relies on.

That said, this is common-moonmage.lic, not common.lic so I'd guess chances are low that it's going to cause problems.

jayRyan24 commented:
Yup, that's it! Multiple chars! Anywho, I would like to just move this in at this point! If not, it's fine, I'm just bored!

rpherbig commented:
I'll merge it tomorrow

Tarhish commented:
Since I'm poking around anyway yeah, this and the common change look like it should work out to me, so thanks for adding it!

I have a tiny suggestion: long term I think the base.yaml entry for telescope_storage: should look like divination_bones_storage:
```
divination_bones_storage:
  container:
  tied:
```
I don't think it has any technical impact whatsoever, but it'll let someone looking in base know that's what it's supposed to look like.

jayRyan24 commented:
I saw it but didn't want to add both and make them think they'd have to have them both set. So I left them out. I did add to the repo info on the wiki though to explain the setting. And they will also be able to check the YAMLs for MM's inside the repo.
